### PR TITLE
MGMT-19498:Add configmap support to the accelerator's collector

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
accelerator collector needs to have a configuration that describes which vendors and models it should collect. This PR add loading that data from a file. Eventually this file be a configmap mapped into the node-exporter pod.
Parsing the configmap file is done in the init function. The file is defined by an environment variable. This variable will be configured by CMO and by unit-test

### CODE REVIEW

**Review by danielmellado — CHANGES_REQUESTED** [review]
Hi @yevgeny-shnaidman thanks for your PR! Have you tried to propose a feature upstream rather than submitting this here?

**Comment by yevgeny-shnaidman** [comment]
> Hi @yevgeny-shnaidman thanks for your PR! Have you tried to propose a feature upstream rather than submitting this here?

Hi @danielmellado , this collector (accelerator) is specific for RedHat needs, as is the configuration configmap , so i am not sure if this is a change suitable for upstream. @simonpasquier wdyt?

**Review by simonpasquier — COMMENTED** [review]
I would do it the other way around:
1. Modify CMO to mount the configmap into the node_exporter container
2. This PR

To simplify error handling and be explicit, an invalid/missing mapping file should exit the exporter (we have e2e tests that should prevent this issue hitting the customer clusters).

**Comment by yevgeny-shnaidman** [comment]
> I would do it the other way around:
> 
> 1. Modify CMO to mount the configmap into the node_exporter container
> 2. This PR
> 
> To simplify error handling and be explicit, an invalid/missing mapping file should exit the exporter (we have e2e tests that should prevent this issue hitting the customer clusters).

I added the error handling so that in case init function of the accelerator is failing, at least all the rest of the functionality will be running. So we loose only the accelerators metrics and not all the rest...

File: collector/accelerators.go
Comment (simonpasquier) [inline collector/accelerators.go:156]: can we return an error if we detect duplicate vendor and PCI ids?

**Review by simonpasquier — COMMENTED** [review]
Unrelated to this PR but I realize that CMO won't redeploy automatically the node_exporter daemonset if the accelerator config file has changed. The usual way would be to hash the content of the config and store it as an annotation in the pod template. Given that CMO doesn't modify the configmap content, the hash can be directly stored in the daemonset manifest in the assets/ directory (using the std.sha1() function, https://jsonnet.org/ref/stdlib.html).

File: collector/accelerators_test.go
Comment (machine424) [inline collector/accelerators_test.go:48]: I think this is better done in a new test where we call `prepareVendorModelData` directly instead of simulating a kingpin parsing.
(CMO e2e tests ensure that parsing is done correctly in the prod code)

File: collector/accelerators.go
Comment (machine424) [reply collector/accelerators.go:156]: Using `yaml.UnmarshalStrict` could be helpful in other cases as well.

File: collector/accelerators.go
Comment (machine424) [inline collector/accelerators.go:57]:
```suggestion
	mappingFile = kingpin.Flag("collector.accelerators.mapping-file", "Path to the mapped accelerators data config.").Default(
		"/var/node_exporter/accelerators_collector_config/config.yaml").String()
```

or sth else.

**Review by machine424 — COMMENTED** [review]
some suggestions

**Comment by machine424** [comment]
> Unrelated to this PR but I realize that CMO won't redeploy automatically the node_exporter daemonset if the accelerator config file has changed. The usual way would be to hash the content of the config and store it as an annotation in the pod template. Given that CMO doesn't modify the configmap content, the hash can be directly stored in the daemonset manifest in the assets/ directory (using the std.sha1() function, https://jsonnet.org/ref/stdlib.html).

I thought about this, but I realized that such change can only occur in a new version which means a different `image` which will lead to the recreation of the Pods after the update of the configmap https://github.com/openshift/cluster-monitoring-operator/blob/a4f514a7bd577e7bb7bbc1405241690028661a0b/pkg/tasks/nodeexporter.go#L97-L111

But if we want to be really sure (in case the image SHA stays the same, which should never happen in the current setup), we can add our own hash for sure.

**Comment by yevgeny-shnaidman** [comment]
> Unrelated to this PR but I realize that CMO won't redeploy automatically the node_exporter daemonset if the accelerator config file has changed. The usual way would be to hash the content of the config and store it as an annotation in the pod template. Given that CMO doesn't modify the configmap content, the hash can be directly stored in the daemonset manifest in the assets/ directory (using the std.sha1() function, https://jsonnet.org/ref/stdlib.html).

Since it was agreed that any change to the configmap will be done only in the new version, it means that the change will be propagated to the new node-exporter node

File: collector/accelerators.go
Comment (yevgeny-shnaidman) [reply collector/accelerators.go:156]: @machine424 which other cases are you referring to? We only have to unmarshal the configmap once. Or fo you mean to use UnmarshalStrict instread of Unmarshal?

Comment (machine424) [reply collector/accelerators.go:156]: yes, instead of Unmarshal to help with incorrect fields names, malformed yaml etc. In addition to adding to the checks that Simon suggested.

**Comment by simonpasquier** [comment]
> Since it was agreed that any change to the configmap will be done only in the new version, it means that the change will be propagated to the new node-exporter node

:facepalm: of course! Forget what I said then.

File: collector/accelerators.go
Comment (yevgeny-shnaidman) [reply collector/accelerators.go]: fixed

File: collector/accelerators_test.go
Comment (yevgeny-shnaidman) [reply collector/accelerators_test.go]: fixed

File: collector/accelerators.go
Comment (machine424) [inline collector/accelerators.go:158]:
```suggestion
			return nil, fmt.Errorf("mapping file contains duplicate of vendor id %s", vendorModelsConfig.VendorID)
```

File: collector/accelerators.go
Comment (machine424) [inline collector/accelerators.go:163]:
```suggestion
				return nil, fmt.Errorf("mapping file contains duplicate of device id %s for vendor id %s", model.PciID, vendorModelsConfig.VendorID)
```

File: collector/accelerators_test.go
Comment (machine424) [inline collector/accelerators_test.go:79]:
```suggestion
	_, err := prepareVendorModelData("accelerators_test_data_duplicated_vendors.bad.yaml")
```

File: collector/accelerators_test.go
Comment (machine424) [inline collector/accelerators_test.go:84]:
```suggestion
	_, err = prepareVendorModelData("testdata/accelerators_test_data_duplicated_device_ids.bad.yaml")
```

To make it clearer the config loading should fail.

File: collector/accelerators_test.go
Comment (machine424) [inline collector/accelerators_test.go:78]:
```suggestion
func Test_prepareVendorModelData_badMapping(t *testing.T) {
```

**Review by machine424 — COMMENTED** [review]
just some nits.
/lgtm
/hold
to wait for the others a little bit.
Thanks for taking care of this.

File: collector/accelerators_test.go
Comment (yevgeny-shnaidman) [reply collector/accelerators_test.go]: fixed

File: collector/accelerators_test.go
Comment (yevgeny-shnaidman) [reply collector/accelerators_test.go]: fixed

File: collector/accelerators_test.go
Comment (yevgeny-shnaidman) [reply collector/accelerators_test.go]: fixed

File: collector/accelerators.go
Comment (yevgeny-shnaidman) [reply collector/accelerators.go]: fixed

File: collector/accelerators.go
Comment (yevgeny-shnaidman) [reply collector/accelerators.go]: fixed

**Comment by machine424** [comment]
Some of the failures seem legit, from one the node-exporter pods

```
        message: "ts=2024-12-17T10:39:54.123Z caller=node_exporter.go:193 level=info
          msg=\"Starting node_exporter\" version=\"(version=1.8.2, branch=master,
          revision=b47bcd0e7907a8a218d0e6fd04ccf17aaa084066)\"\nts=2024-12-17T10:39:54.123Z
          caller=node_exporter.go:194 level=info msg=\"Build context\" build_context=\"(go=go1.23.2
          (Red Hat 1.23.2-1.el9) X:strictfipsruntime, platform=linux/amd64, user=root@ef450955865c,
          date=20241217-06:22:40, tags=strictfipsruntime)\"\nts=2024-12-17T10:39:54.123Z
          caller=diskstats_common.go:111 level=info collector=diskstats msg=\"Parsed
          flag --collector.diskstats.device-exclude\" flag=^(z?ram|loop|fd|(h|s|v|xv)d[a-z]|nvme\\d+n\\d+p)\\d+$\npanic:
          Couldn't create metrics handler: couldn't create collector: failed to get
          the accelerator configuration: failed to unmarshal accelerators config data:
          yaml: unmarshal errors:\n\t  line 3: field pciName not found in type struct
          { PciID string \"yaml:\\\"pciID\\\"\"; ModelName string \"yaml:\\\"modelName\\\"\"
          }\n\t  line 5: field pciName not found in type struct { PciID string \"yaml:\\\"pciID\\\"\";
          ModelName string \"yaml:\\\"modelName\\\"\" }\n\t  line 7: field pciName
          not found in type struct { PciID string \"yaml:\\\"pciID\\\"\"; ModelName
          string \"yaml:\\\"modelName\\\"\" }\n\t  line 9: field pciName not found
          in type struct { PciID string \"yaml:\\\"pciID\\\"\"; ModelName string \"yaml:\\\"modelName\\\"\"
          }\n\t  line 11: field pciName not found in type struct { PciID string \"yaml:\\\"pciID\\\"\";
          ModelName string \"yaml:\\\"modelName\\\"\" }\n\t  line 13: field pciName
          not found in type struct { PciID string \"yaml:\\\"pciID\\\"\"; ModelName
          string \"yaml:\\\"modelName\\\"\" }\n\t  line 15: field pciName not found
          in type struct { PciID string \"yaml:\\\"pciID\\\"\"; ModelName string \"yaml:\\\"modelName\\\"\"
          }\n\t  line 17: field pciName not found in type struct { PciID string \"yaml:\\\"pciID\\\"\";
          ModelName string \"yaml:\\\"modelName\\\"\" }\n\ngoroutine 1 [running]:\nmain.newHandler(0x1,
          0x28, {0xd802a0, 0xc000090200})\n\t/go/src/github.com/prometheus/node_exporter/node_exporter.go:69
          +0x2a5\nmain.main()\n\t/go/src/github.com/prometheus/node_exporter/node_exporter.go:201
          +0x128d\n"
```

It seems the `pciName` in use on CMO side now is unknown, we should fix that first. or adjust the PR

**Comment by machine424** [comment]
Let's run the failing tests with that PR to be sure it's enough


/testwith openshift/node-exporter/master/e2e-aws openshift/cluster-monitoring-operator#2540
/testwith openshift/node-exporter/master/e2e-aws-upgrade openshift/cluster-monitoring-operator#2540

**Comment by machine424** [comment]
/testwith openshift/node-exporter/master/e2e-aws openshift/node_exporter#155 openshift/cluster-monitoring-operator#2540
/testwith openshift/node-exporter/master/e2e-aws-upgrade openshift/node_exporter#155 openshift/cluster-monitoring-operator#2540

**Comment by yevgeny-shnaidman** [comment]
/testwith openshift/node-exporter/master/e2e-aws https://github.com/openshift/node_exporter/pull/155 https://github.com/openshift/cluster-monitoring-operator/pull/2540

**Comment by machine424** [comment]
(they were triggered, `node_exporter` instead of `node-exporter`)

**Comment by machine424** [comment]
/lgtm
once the CMO PR is merged, this should be mergeable.

**Comment by yevgeny-shnaidman** [comment]
/retitle MGMT-19498:Add configmap support to the accelerator's collector

**Comment by machine424** [comment]
Not sure what's blocking this but
/lgtm
again, in case.

**Comment by machine424** [comment]
/lgtm
I assume this can wait until `acknowledge-critical-fixes-only` is lifted?

**Comment by yevgeny-shnaidman** [comment]
> /lgtm I assume this can wait until `acknowledge-critical-fixes-only` is lifted?

We can push it now, if you are ok with it, it is for 4.19 only

**Comment by machine424** [comment]
let's do it
/label acknowledge-critical-fixes-only
I'll be keep an eye on the builds.
(reverting is always an option)

**Pull request merged**
